node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+								let job_data = &mut tracker.per_relay_parent.get_mut(&hash).expect("Overseer does not send work items related to relay parents that are not part of our workset. qed");
+
+								let validator = {
+									job_data
+										.validator_set
+										.get(signed_availability.validator_index() as usize)
+										.expect("Our own validation index exists. qed")
+								}
+								.clone();
+
+								let peer_views = &mut tracker.peer_views;
+								let msg = BitfieldGossipMessage {
+									relay_parent: hash,
+									signed_availability,
+								};
+
+								relay_message(&mut ctx, job_data, peer_views, validator, msg)
+									.await?;
+							}
+							BitfieldDistributionMessage::NetworkBridgeUpdate(event) => {
+								trace!(target: "bitd", "Processing NetworkMessage");
+								// a network message was received
+								if let Err(e) =
+									handle_network_msg(&mut ctx, &mut tracker, event).await
+								{
+									warn!(target: "bitd", "Failed to handle incomming network messages: {:?}", e);
+								}
+							}
+						}
+					}
+					FromOverseer::Signal(OverseerSignal::StartWork(relay_parent)) => {
+						trace!(target: "bitd", "Start {:?}", relay_parent);
+						// query basic system parameters once
+						// @todo assumption: these cannot change within a session
+						let (validator_set, signing_context) =
+							query_basics(&mut ctx, relay_parent).await?;
+
+						let _ = tracker.per_relay_parent.insert(
+							relay_parent,
+							PerRelayParentData {
+								signing_context,
+								validator_set,
+								..Default::default()
+							},
+						);
+					}
+					FromOverseer::Signal(OverseerSignal::StopWork(relay_parent)) => {
+						trace!(target: "bitd", "Stop {:?}", relay_parent);
+						// @todo assumption: it is good enough to prevent additional work from being
+						// scheduled, the individual futures are supposedly completed quickly
+						let _ = tracker.per_relay_parent.remove(&relay_parent);
+					}
+					FromOverseer::Signal(OverseerSignal::Conclude) => {
+						trace!(target: "bitd", "Conclude");
+						tracker.per_relay_parent.clear();
+						return Ok(());
+					}
+				}
+			}
+		}
+	}
+}
+
+/// Modify the reputation of peer based on their behaviour.
+async fn modify_reputation<Context>(
+	ctx: &mut Context,
+	peer: PeerId,
+	rep: ReputationChange,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	trace!(target: "bitd", "Reputation change of {:?} for peer {:?}", rep, peer);
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::ReportPeer(peer, rep),
+	))
+	.await
+}
+
+/// Distribute a given valid and signature checked bitfield message.
+///
+/// Can be originated by another subsystem or received via network from another peer.
+async fn relay_message<Context>(
+	ctx: &mut Context,
+	job_data: &mut PerRelayParentData,
+	peer_views: &mut HashMap<PeerId, View>,
+	validator: ValidatorId,
```

Comment:
The `validator` here is  redundant if `message_sent_to_peer` is altered to hold `ValidatorIndex` and not `ValidatorId`. There is a one-to-one mapping between them for any given relay-parent. And the validator index is present within the signed bitfield itself.

Reply:
Redundant yes, but from where the message is used, the validator is already extracted and hence error handling is externalised from this `fn`.